### PR TITLE
fix: handling zero value responses (previously undefined)

### LINKED ISSUE #165: [flagd-provider] add unit test for zero-value resolutions from flagd
Flagd sends empty values in flag [response payloads](https://github.com/open-feature/schemas/blob/main/protobuf/schema/v1/schema.proto#L54) when the value is equivalent to the zero-value of the associated go type. We should test that this works as expected and still returns an equivalent zero-value (ie: `false`, `0`).

DoD:
- manual test confirming zero values are returned from the flagd provider when they are missing from the gRPC message (boolean set to false or numeric set to 0 in flagd config)
- new tests that assert correct values for zero-value flags

see: https://github.com/open-feature/flagd/issues/218

### PR DESCRIPTION
**Pull request by odubajDT**
## This PR

- introduce logic for zero value handling
- unit tests

### Related Issues

Fixes #165

### CODE REVIEW

**Review by Kavindu-Dodan — APPROVED** [review]
lgtm

File: libs/providers/flagd/src/lib/service/grpc/service.ts [hunk 1]
```diff
@@ -206,17 +206,37 @@ export class GRPCService implements Service {
   private objectParser = (struct: Struct) => {
     if (struct !== undefined) {
```
Comment (thisthat) [inline libs/providers/flagd/src/lib/service/grpc/service.ts:207]: maybe it is worth to also nullcheck here
```suggestion
    if (struct) {
```

**Review by james-milligan — APPROVED** [review]
LGTM

**Review by thisthat — APPROVED** [review]

File: libs/providers/flagd/src/lib/service/grpc/service.ts [hunk 2]
```diff
@@ -204,19 +204,39 @@ export class GRPCService implements Service {
...
+    return 0
   };
 
   private async resolve<T extends FlagValue, Rq extends AnyRequest, Rs extends AnyResponse>(
     resolver: (request: AnyRequest) => UnaryCall<Rq, Rs>,
     flagKey: string,
     context: EvaluationContext,
     logger: Logger,
-    parser?: (struct: Struct) => PbJsonValue
+    parser?: (value: any) => any
```
Comment (toddbaert) [inline libs/providers/flagd/src/lib/service/grpc/service.ts:239]: Can we restrict these types a bit more?

It can be a union of types: `Struct | number | string...`. Generally, we shouldn't see explicit `any`s (`unknown` is safer if you really don't know the type)

File: libs/providers/flagd/src/lib/service/grpc/service.ts [hunk 3]
```diff
@@ -114,15 +114,15 @@ export class GRPCService implements Service {
...
   }
 
   async resolveString(flagKey: string, context: EvaluationContext, logger: Logger): Promise<ResolutionDetails<string>> {
-    return this.resolve(this._client.resolveString, flagKey, context, logger);
+    return this.resolve(this._client.resolveString, flagKey, context, logger, this.stringParser);
   }
 
   async resolveNumber(flagKey: string, context: EvaluationContext, logger: Logger): Promise<ResolutionDetails<number>> {
-    return this.resolve(this._client.resolveFloat, flagKey, context, logger);
+    return this.resolve(this._client.resolveFloat, flagKey, context, logger, this.numberParser);
```
Comment (toddbaert) [inline libs/providers/flagd/src/lib/service/grpc/service.ts:125]: So it looks like this wasn't working as expected? When a zero-value flag was set in flagd config, `undefined` was returned from the provider? Did you see that in a manual test?

Comment (odubajDT) [reply libs/providers/flagd/src/lib/service/grpc/service.ts:125]: yes exactly, this logic was not in place

File: libs/providers/flagd/src/lib/service/grpc/service.ts (hunk 2 above)
Comment (odubajDT) [reply libs/providers/flagd/src/lib/service/grpc/service.ts:239]: it was quite problematic here, but will give it another try. If it won't go, I will use `undefined`

Comment (odubajDT) [reply libs/providers/flagd/src/lib/service/grpc/service.ts:239]: I am trying desperately but cannot manage to make it work without the use of `any`. The problem is the conversion of the object resolution, which in this case can be anything. Would you maybe have time to sync regarding this problem? I do not have that deep knowledge regarding JS/Typescript, so maybe I am missing some good manners how to resolve this here...

Comment (thisthat) [reply libs/providers/flagd/src/lib/service/grpc/service.ts:239]: The problem here is that `Struct` and `FlagValue` are not compatible types.

Comment (odubajDT) [reply libs/providers/flagd/src/lib/service/grpc/service.ts:239]: we were working on this together with @thisthat but didn't manage to find a solution. For now we prefer to stay with the current solution

File: libs/providers/flagd/src/lib/service/grpc/service.ts (hunk 3 above)
Comment (toddbaert) [reply libs/providers/flagd/src/lib/service/grpc/service.ts:125]: Perfect! This was a productive issue then.

Thanks @odubajDT and @james-milligan
